File: github_cpp/12/29.cpp
    #include <bits/stdc++.h>
     
    using namespace std;
     
    struct node {
        int weight;
        pair<int, int> edge;
    };
     
    struct ds {
        int val;
        int size;
    };
     
    bool comparator(const node& node1, const node& node2) {
        return node1.weight < node2.weight;
    }
     
    vector<ds> Arr;
     
    void initialize(int n) {
        for(int i = 0; i < n; ++i) {
            ds temp;
            temp.val = i;
            temp.size = 1;
     
            Arr.push_back(temp);
        }
    }
     
    int root(int i) {
        while(Arr.at(i).val != i) {
            Arr.at(i).val = Arr.at(Arr.at(i).val).val;
            i = Arr.at(i).val;
        }
     
        return i;
    }
     
    void WeightedUnion(int a, int b) {
        int r_a = root(a);
        int r_b = root(b);
     
        if(r_a == r_b) 
            return;
     
        if(Arr.at(r_a).size < Arr.at(r_b).size) {
            Arr.at(r_a).val = Arr.at(r_b).val;
            Arr.at(r_b).size +=  Arr.at(r_a).size;
        }
        else {
            Arr.at(r_b).val = Arr.at(r_a).val;
            Arr.at(r_a).size +=  Arr.at(r_b).size;
        }
        
    }
     
    int Find(int a, int b) {
        return root(a) == root(b);
    }
     
    int Kruskal(vector<node>& Graph) {
        int minCost = 0;
     
        for(auto& i : Graph) {
            
            if(!Find(i.edge.first, i.edge.second)) {
                WeightedUnion(i.edge.first, i.edge.second);
                minCost += i.weight;
            }    
        }
     
        return minCost;
    }
     
    int main() {
        int nodes, edges;
        cin>>nodes>>edges;
     
        vector<node> Graph(edges+1);
        initialize(nodes+1);
     
        for(int i = 1; i <= edges; ++i) {
            int x, y, w;
            cin>>x>>y>>w;
     
            Graph.at(i).weight = w;
            Graph.at(i).edge = make_pair(x, y);
        }
     
        sort(Graph.begin(), Graph.end(), comparator);
     
        cout<<Kruskal(Graph)<<endl;
        
        return 0;
    }

File: github_code/shell-sort/java/29.java
package algorithm.sort;

public class ShellSort {
    /*private void shellsort(int[] a) {
        int d = a.length / 2;
        while (true) {
            for (int i = 0; i < d; i++) {
                for (int j = i; j + d < a.length; j += d) {
                    int temp;
                    if (a[j] > a[j + d]) {
                        temp = a[j];
                        a[j] = a[j + d];
                        a[j + d] = temp;
                    }
                }
            }
            if (d == 1) {
                break;
            }
            d--;
        }
    }*/

    private void shellsort(int[] a) {
        int d = a.length / 2;
        while (true) {
            for (int i = 0; i < a.length; i++) {
                for (int j = i; j + d < a.length; j += d) {
                    int temp;
                    if (a[j] > a[j + d]) {
                        temp = a[j];
                        a[j] = a[j + d];
                        a[j + d] = temp;
                    }
                }
            }
            d--;
            if (d == 1) {
                break;
            }
        }
    }


    public static void main(String[] args) {
        int[] a = {11, 21, 42, 62, 1, 4, 2, 52, 73, 36, 98};
        ShellSort shellSort = new ShellSort();
        //shellSort.shellsort(a);
        for (int arr : a) {
            System.out.println(arr);
        }
    }
}
